ring/ingester_lifecycle.go

```diff
@@ -16,125 +16,214 @@
 	defer ticker.Stop()
 	for {
 		select {
 		case <-ticker.C:
-			log.Info("Adding ingester to consul")
-			if err := r.consul.PutBytes(r.id, r.desc); err != nil {
+			log.Infof("Heartbeating to consul...")
+			if err := r.consul.CAS(ring, descFactory, func(in interface{}) (out interface{}, retry bool, err error) {
+				ringDesc := &Desc{}
+				if in != nil {
+					ringDesc = in.(*Desc)
+				}
+
+				ingesterDesc, ok := ringDesc.Ingesters[r.id]
+				if !ok {
+					// consul must have restarted
+					log.Infof("Found empty ring, inserting tokens!")
+					populateRingDesc(ringDesc, r.id, r.hostname, tokens)
+				} else {
+					ingesterDesc.Timestamp = time.Now()
+					ringDesc.Ingesters[r.id] = ingesterDesc
+				}
+
+				return ringDesc, true, nil
+			}); err != nil {
 				log.Errorf("Failed to write to consul, sleeping: %v", err)
-			} else {
-				return
 			}
 		case <-r.quit:
 			return
 		}
 	}
 }
 
-// describeLocalIngester returns an IngesterDesc for the ingester that is this
-// process.
-func describeLocalIngester(listenPort, numTokens int) (*IngesterDesc, error) {
-	hostname, err := os.Hostname()
-	if err != nil {
-		return nil, err
-	}
+// Unregister deletes ingestor config from Consul
+func (r *IngesterRegistration) Unregister() {
+	log.Info("Removing ingester from consul")
+	close(r.quit)
+	r.wait.Wait()
+}
 
-	addr, err := getFirstAddressOf(infName)
-	if err != nil {
-		return nil, err
-	}
+func (r *IngesterRegistration) unregister(tokens []uint32) {
+	if err := r.consul.CAS(ring, descFactory, func(in interface{}) (out interface{}, retry bool, err error) {
+		if in == nil {
+			log.Error("Found empty ring when trying to unregister!")
+			return nil, false, nil
+		}
 
-	return &IngesterDesc{
-		ID:       hostname,
-		Hostname: fmt.Sprintf("%s:%d", addr, listenPort),
-		Tokens:   generateTokens(hostname, numTokens),
-	}, nil
+		ringDesc := in.(*Desc)
+		removeFromRingDesc(ringDesc, r.id, tokens)
+		return ringDesc, true, nil
+	}); err != nil {
+		log.Fatalf("Failed to unregister from consul: %v", err)
+	}
 }
 
-func generateTokens(id string, numTokens int) []uint32 {
-	tokenHasher := fnv.New64()
-	tokenHasher.Write([]byte(id))
-	r := rand.New(rand.NewSource(int64(tokenHasher.Sum64())))
-
-	tokens := []uint32{}
-	for i := 0; i < numTokens; i++ {
-		tokens = append(tokens, r.Uint32())
+type sortableUint32 []uint32
+
+func (ts sortableUint32) Len() int           { return len(ts) }
+func (ts sortableUint32) Swap(i, j int)      { ts[i], ts[j] = ts[j], ts[i] }
+func (ts sortableUint32) Less(i, j int) bool { return ts[i] < ts[j] }
+
+func generateTokens(numTokens int, takenTokens []uint32) []uint32 {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	tokens := sortableUint32{}
+	for i := 0; i < numTokens; {
+		candidate := r.Uint32()
+		j := sort.Search(len(takenTokens), func(i int) bool {
+			return takenTokens[i] >= candidate
+		})
+		if j < len(takenTokens) && takenTokens[j] == candidate {
+			continue
+		}
```

Comment:
Just checking: these 6 lines of code are here to implement:

```python
if candidate in takenTokens:
    continue
```

But with a binary search?